contextual_lenses/train_utils.py

```diff
@@ -0,0 +1,136 @@
+"""Train utils
+
+General tools for instantiating and training models."""
+
+
+import flax
+from flax import nn
+from flax import optim
+from flax.training import checkpoints
+from flax.training import common_utils
+
+import jax
+from jax import random
+import jax.nn
+import jax.numpy as jnp
+
+import numpy as np
+
+import functools
+
+
+def create_optimizer(model, learning_rate, weight_decay):
+  """Instantiates Adam optimizer."""
+
+  optimizer_def = optim.Adam(learning_rate=learning_rate, weight_decay=weight_decay)
+  optimizer = optimizer_def.create(model)
+  
+  return optimizer
+
+
+@functools.partial(jax.jit, static_argnums=(3, 4))
+def train_step(optimizer, X, Y, loss_fn, loss_fn_kwargs):
+  """Trains model (optimizer.target) using specified loss function."""
+
+  def compute_loss_fn(model, X, Y, loss_fn, loss_fn_kwargs):
+    Y_hat = model(X)
+    loss = loss_fn(Y, Y_hat, **loss_fn_kwargs)
+    return loss
+  
+  grad_fn = jax.value_and_grad(compute_loss_fn)
+  _, grad = grad_fn(optimizer.target, X, Y, loss_fn, loss_fn_kwargs)
+  optimizer = optimizer.apply_gradient(grad)
+  
+  return optimizer
+
+
+def train(model, train_data, loss_fn, loss_fn_kwargs, learning_rate=1e-4, weight_decay=0.1,
```

Comment:
you should have a test somewhere that if you write state to disk and load back in you get the same thing back. Especially important when restoring optimizers.